breaking change: credential_pop is an array (#36)
peppelinux/draft-demarco-oauth-status-attestations: draft-demarco-oauth-status-attestations.md

+--------+----------+                         +----------+---------+
         |                                               |
         | HTTP POST /status                             |

>          |  credential_pop = [$CredentialPoPJWT]         |

Reviewer: since the value is an array, wouldn't be better to rename this to `credentials_pop`?

@marinaado 
@fmarino-ipzs 
@OR13

Reviewer 2: I wonder if these are all really "credentials"... or are they simply assertions related to a credential that had a confirmation claim?

- https://datatracker.ietf.org/doc/html/rfc7800#section-3.1
- https://datatracker.ietf.org/doc/html/rfc8747#section-3.1

Now that its an array, can each of these items demonstrate possession of a different key?

Can methods be mixed, for example, cred-1 uses COSE Key, cred-2 use JKT ?

Reviewer: Given the updates from version 00 to 01, the term credentials is misleading because the items within that array are not credentials but assertions that prove credential possession. Therefore, we should rename this parameter to better reflect its purpose.

Reviewer: another solution @SaraConsoliACN @OR13 to prevent the use of `application/json` would be this

`credential_pop="..."&credential_pop="..."&...`

taking into account that credential_pop is deprecable now and require a name change

Reviewer 2: If the data posted is not credentials, it might be better to give it a more generic name, like `presentations` or `arguments`, or `parameters`

Reviewer: we have a section defining the Digital Credential PoP here: https://github.com/peppelinux/draft-demarco-oauth-status-attestations/pull/36/files#diff-6018989306b8996e79d471eb123fc3b59796d1635384b6faac8cdb36cd5f49ccR297

I'm not sure about the name presentation

@SaraConsoliACN please revert the change to credential_pop to allow this PR to be merged and create another issue for discussing the json member name change

Reviewer: @OR13 the discussion is moved to this issue: https://github.com/peppelinux/draft-demarco-oauth-status-attestations/issues/43

I'd suggest to unblock this PR and continue the discussion about the nature of the parameters and the good semantic to have outside of this PR if we agree

@fmarino-ipzs @marinaado ^

~~~
POST /status HTTP/1.1
Host: issuer.example.org
Content-Type: application/json

> { "credential_pop" : [$CredentialPoPJWT, $CredentialPoPCWT, ... ] }

Reviewer 2: if these need more params, we might consider this interface to be an array of objects containing encoded credentials.

Reviewer: What about encoding any required parameter within the JWT/CWT according to the schema type `status-attestation+{jwt,cwt}`?

this would make each object consistent in its purpose

we may say that any additional parameter could be included, while only the required set of attributes, for each status request is defined within the specs

Reviewer: therefore we could rename credential_pop to `status_assertion_requests` where each element within the array configures a `typ` value set to `status-assertion-request+{jwt,cwt}`

the corresponding response would be a json array containing for each element with a `typ` value set to `status-assertion+{jwt,cwt}` or ar `status-assertion-error+{jwt,cwt}` at the same position of the corresponding one within the request.

Reviewer 2: You would need to verify / decode the protected header to distinguish the types... but it is nice that they are required to advertise their typ consistent with the JWT BCP.

Reviewer: yes, I was wondering about "in object veritas"

Host: issuer.example.org

> Content-Type: application/x-www-form-urlencoded

Reviewer 2: I wonder if this should be changed to JSON.

If it says like this, I think an example showing application/x-www-form-urlencoded of CWT and JWT would help communicate the intention here, I am happy to produce one if there is agreement.

Reviewer: I'm also in favor of `application/json` too

@marinaado @fmarino-ipzs ^

~~~ ascii-art
+-------------------+                                  +--------------------+
|                   |                                  |                    |
|  Wallet Instance  |                                  | Credential Issuer  |
|                   |                                  |                    |
+--------+----------+                                  +----------+---------+
         |                                                        |
         | HTTP POST /status                                      |
         |  status_assertion_requests = [$StatusAssertionRequest] |
         +-------------------------------------------------------->
         |                                                        |

>          |  Response with Status Assertion                        |

Reviewer: ```suggestion
         |  Status Assertion Responses [...]                      |
```

to the confirmation claim assigned by the issuer and contained within
the Digital Credential.

Below a non-normative example representing a Status Assertion Request array with a
single JWT in it.

~~~
POST /status HTTP/1.1
Host: issuer.example.org
Content-Type: application/json

"status_assertion_requests" : ["${base64url(json({typ: (some pop for status-assertion)+jwt, ...}))}.payload.signature", ... ]
~~~

> The Status Assertion HTTP request can be sent to a single Credential Issuer regarding multiple Digital Credentials, and MUST contain a json object with the member `status_assertion_requests`.

Reviewer: ```suggestion
The Status Assertion HTTP request can be sent to a single Credential Issuer
regarding multiple Digital Credentials, and MUST contain a json object with
the member `status_assertion_requests`.
```

The `status_assertion_requests` MUST be set with an array of strings.

Each string within the array is a Digital Credential Status Assertion Request.

> The Credential Issuer that receives the Status Assertion Request MUST validate that the Wallet Instance making the request is authorized to request Status Assertions.

Reviewer: ```suggestion
The Credential Issuer that receives the Status Assertion Request
MUST validate that the Wallet Instance making the request is
authorized to request Status Assertions.
```

to the confirmation claim assigned by the issuer and contained within
the Digital Credential.

Below a non-normative example representing a Status Assertion Request array with a
single JWT in it.

~~~
POST /status HTTP/1.1
Host: issuer.example.org
Content-Type: application/json

"status_assertion_requests" : ["${base64url(json({typ: (some pop for status-assertion)+jwt, ...}))}.payload.signature", ... ]
~~~

Given that the Wallet may request one or more Status Attestations from the same Credential Issuer,
the `status_assertion_requests` parameter is subject to the following specification:
- Status_assertion_requests: REQUIRED. It MUST be implemented as an array of strings,
where each string represents a Digital Credential proof of possession
- The position of each `$StatusAssertionRequest` object
within the array MUST match the same position in the subsequent response.

To validate that the Wallet Instance is entitled to request its Status Assertion,

> the following requirements MUST be satisfied:

Reviewer: ```suggestion
Therefore the following requirements MUST be satisfied:
```

- The Credential Issuer MUST verify the signature of all elements in the `status_assertion_requests` object

> using the public key contained within the Digital Credential where the `status_assertion_requests` is referred to;

Reviewer: please use the term confirmation method instead of public key

to the confirmation claim assigned by the issuer and contained within
the Digital Credential.

Below a non-normative example representing a Status Assertion Request array with a
single JWT in it.

~~~
POST /status HTTP/1.1
Host: issuer.example.org
Content-Type: application/json

"status_assertion_requests" : ["${base64url(json({typ: (some pop for status-assertion)+jwt, ...}))}.payload.signature", ... ]
~~~

The Status Assertion HTTP request can be sent to a single Credential Issuer

> regarding multiple Digital Credentials, and MUST contain a json object with

Reviewer: ```suggestion
regarding multiple Digital Credentials, and MUST contain a JSON object with
```

the member `status_assertion_requests`.

The `status_assertion_requests` MUST be set with an array of strings, where
each string within the array represents a Digital Credential Status Assertion Request.

The Credential Issuer that receives the Status Assertion Request
MUST validate that the Wallet Instance making the request is
authorized to request Status Assertions.
Therefore the following requirements MUST be satisfied:

- The Credential Issuer MUST verify the signature of all elements in the `status_assertion_requests` object

> using the confirmation method contained within the Digital Credential where the `status_assertion_requests` is referred to;

Reviewer: ```suggestion
using the confirmation method contained within the Digital Credential where the Status Assertion Request object is referred to;
```

- The Credential Issuer MUST verify that it is the legitimate Issuer of the Digital Credential

> to which the `status_assertion_requests` refers.

Reviewer: ```suggestion
to which each Status Assertion Request object refers.
```

to the confirmation claim assigned by the issuer and contained within
the Digital Credential.

Below a non-normative example representing a Status Assertion Request array with a
single JWT in it.

~~~
POST /status HTTP/1.1
Host: issuer.example.org
Content-Type: application/json

"status_assertion_requests" : ["${base64url(json({typ: (some pop for status-assertion)+jwt, ...}))}.payload.signature", ... ]
~~~

The Status Assertion HTTP request can be sent to a single Credential Issuer
regarding multiple Digital Credentials, and MUST contain a json object with
the member `status_assertion_requests`.

> The `status_assertion_requests` MUST be set with an array of strings.
> 
> Each string within the array is a Digital Credential Status Assertion Request.

Reviewer: ```suggestion
The `status_assertion_requests` MUST be set with an array of strings, where
each string within the array represents a Digital Credential Status Assertion Request.
```

> ~~~

Reviewer: ```suggestion

Below a non-normative example representing a Status Assertion Request array with a
single JWT in it.

~~~
```

POST /status HTTP/1.1
Host: issuer.example.org
Content-Type: application/json

"status_assertion_requests" : ["${base64url(json({typ: (some pop for status-assertion)+jwt, ...}))}.payload.signature", ... ]
~~~

> Given that the Wallet may request one or more Status Attestations from the same Credential Issuer, 
> the `status_assertion_requests` parameter is subject to the following specification:
> - Status_assertion_requests: REQUIRED. It MUST be implemented as an array of strings,

Reviewer: ```suggestion
The Status Assertion HTTP request can be sent to a single Credential Issuer regardng muliple Digital Credentials, and MUST contain a json object with the member `status_assertion_requests`. 
The `status_assertion_requests` MUST be set with an array of strings,
```

> where each string represents a Digital Credential proof of possession

Reviewer: ```suggestion
Each string within the array is a Digital Credential Status Assertion Request, where

> - The position of each `$StatusAssertionRequest` object

Reviewer: ```suggestion
the position of each `$StatusAssertionRequest` object
```

> - The position of each `$StatusAssertionRequest` object
> within the array MUST match the same position in the subsequent response.

Reviewer: ```suggestion
```

this is related to the response, not to the request. Useless saying this here.

| --- | --- | --- |
| **iss** | Wallet identifier. | {{RFC9126}}, {{RFC7519}} |
| **aud** | It MUST be set with the Credential Issuer Status Assertion endpoint URL as value that identify the intended audience | {{RFC9126}}, {{RFC7519}} |

> | **exp** | UNIX Timestamp with the expiration time of the JWT. It MUST be superior to the value set for `iat`  . | {{RFC9126}}, {{RFC7519}}, {{RFC7515}} |

Reviewer 2: ```suggestion
| **exp** | UNIX Timestamp with the expiration time of the JWT. It MUST be greater than the value set for `iat`  . | {{RFC9126}}, {{RFC7519}}, {{RFC7515}} |
```

  Content-Type: application/json;charset=UTF-8

  {
    "error": "invalid_request"

>     "error_description": "The signature of credential_pop JWT is not valid"

Reviewer: ```suggestion
    "error_description": "The signature of the status assertion request is not valid"
```

| --- | --- | --- |
| **iss** | It MUST be set to the identifier of the Issuer. | {{RFC9126}}, {{RFC7519}} |
| **iat** | UNIX Timestamp with the time of the Status Assertion issuance. | {{RFC9126}}, {{RFC7519}} |

> | **exp** | UNIX Timestamp with the expiration time of the JWT. It MUST be superior to the value set for `iat`. | {{RFC9126}}, {{RFC7519}}, {{RFC7515}} |

Reviewer 2: ```suggestion
| **exp** | UNIX Timestamp with the expiration time of the JWT. It MUST be greater than the value set for `iat`. | {{RFC9126}}, {{RFC7519}}, {{RFC7515}} |
```

| --- | --- | --- |
| **iss** | It MUST be set to the identifier of the Issuer. | {{RFC9126}}, {{RFC7519}} |
| **iat** | UNIX Timestamp with the time of the Status Attestation issuance. | {{RFC9126}}, {{RFC7519}} |

> | **exp** | UNIX Timestamp with the expiration time of the JWT. It MUST be superior to the value set for `iat`  . | {{RFC9126}}, {{RFC7519}}, {{RFC7515}} |

Reviewer: ```suggestion
| **exp** | UNIX Timestamp with the expiration time of the JWT. It MUST be superior to the value set for `iat`. | {{RFC9126}}, {{RFC7519}}, {{RFC7515}} |
```

If the Status Attestation is requested for a non-existent, expired, revoked or invalid Digital Credential, the
Credential Issuer MUST respond with an HTTP Response with the status code set to 404.

> If the Digital Credential is valid, the Credential Issuer MUST return an HTTP status code of 201 (Created), with the content type set to `application/json`. The response MUST include a JSON object with a member named `status_attestations`, which contains a list of Status Attestations, as illustrated in the following non-normative example:

Reviewer: ```suggestion
If the Digital Credential is valid, the Credential Issuer MUST return an HTTP status code of 201 (Created), with the content type set to `application/json`. The response MUST include a JSON object with a member named `status_attestations`, which contains an array of strings. Each string in the array represents a Status Attestation, as illustrated in the following non-normative example:
```

> If the Digital Credential is valid, the Credential Issuer MUST return an HTTP status code of 201 (Created), with the content type set to `application/json`. The response MUST include a JSON object with a member named `status_attestations`, which contains a list of Status Attestation for the Wallet Instance, as illustrated in the following non-normative example:

Reviewer: ```suggestion
If the Digital Credential is valid, the Credential Issuer MUST return an HTTP status code of 201 (Created), with the content type set to `application/json`. The response MUST include a JSON object with a member named `status_attestations`, which contains a list of Status Attestations, as illustrated in the following non-normative example:
```

~~~
HTTP/1.1 201 Created
Content-Type: application/json

{
    "status_attestations": ["eyJhbGciOiJFUzI1Ni ...", "..."],
}
~~~

> The status_attestations object MUST adhere to the following specification:
> - `status_attestations`. REQUIRED. By default, it MUST be an array of strings (array[string]).

Reviewer: ```suggestion

```

| **iss** | Wallet identifier. | {{RFC9126}}, {{RFC7519}} |

> | **aud** | It MUST be set with the Credential Issuer Status Assertion endpoint URL as value that identify the intended audience | {{RFC9126}}, {{RFC7519}} |

Reviewer: ```suggestion
| **aud** | It MUST be set with the Credential Issuer Status Assertion endpoint URL as value that identify the intended audience. | {{RFC9126}}, {{RFC7519}} |
```

| --- | --- | --- |
| **iss** | It MUST be set to the identifier of the Issuer. | {{RFC9126}}, {{RFC7519}} |
| **iat** | UNIX Timestamp with the time of the Status Assertion issuance. | {{RFC9126}}, {{RFC7519}} |
| **exp** | UNIX Timestamp with the expiration time of the JWT. It MUST be greater than the value
set for `iat`. | {{RFC9126}}, {{RFC7519}}, {{RFC7515}} |
| **credential_hash** | Hash value of the Digital Credential the Status Assertion is bound to. | this specification |
| **credential_hash_alg** | The Algorithm used of hashing the Digital Credential to which the Status Assertion is bound.
The value SHOULD be set to `sha-256`. | this specification |

> | **cnf** | JSON object containing the cryptographic key binding. The `cnf.jwk` value MUST match

Reviewer: instead of cryptographic key binding we should use the term confirmation method to enable other bindings mechanisms

| --- | --- | --- |
| **iss** | It MUST be set to the identifier of the Issuer. | {{RFC9126}}, {{RFC7519}} |
| **iat** | UNIX Timestamp with the time of the Status Assertion issuance. | {{RFC9126}}, {{RFC7519}} |
| **exp** | UNIX Timestamp with the expiration time of the JWT. It MUST be greater than the value
set for `iat`. | {{RFC9126}}, {{RFC7519}}, {{RFC7515}} |
| **credential_hash** | Hash value of the Digital Credential the Status Assertion is bound to. | this specification |
| **credential_hash_alg** | The Algorithm used of hashing the Digital Credential to which the Status Assertion is bound.
The value SHOULD be set to `sha-256`. | this specification |

> | **cnf** | JSON object containing the confirmation method. The `cnf.jwk` value MUST match

Reviewer: ```suggestion
| **cnf** | JSON object containing the confirmation method. Its value MUST be compliant with one contained in the related Digital Credential. For instance, if `cnf.jwk` is used within the Digital Credential, the same value MUST set within the Status Assertion Request.
```

If the Digital Credential is valid, the Credential Issuer MUST return
an HTTP status code of 201 (Created), with the content type set to
`application/json`. The response MUST include a JSON object with a member
named `status_assertion_responses`, which contains the Status Assertion for the
Wallet Instance, as illustrated in the following non-normative example:

~~~
HTTP/1.1 201 Created
Content-Type: application/json

{
    "status_assertion_responses": ["${base64url(json({typ: status-assertion+jwt, ...}))}.payload.signature", ... ]
}
~~~

> Given that the member `status_assertion_responses` is an array of strings, it is subject

Reviewer: ```suggestion
The member `status_assertion_responses` MUST be an array of strings, 
where each of them represent a Status Assertion Response object. In particular:
```

If the Digital Credential is valid, the Credential Issuer MUST return
an HTTP status code of 201 (Created), with the content type set to
`application/json`. The response MUST include a JSON object with a member

> named `status_assertion_responses`, which contains the Status Assertion for the

Reviewer: ```suggestion
named `status_assertion_responses`, which contains the Status Assertions and or the Status Assertion Errors related to the request made by the
```

> Wallet Instance, as illustrated in the following non-normative example:

Reviewer: ```suggestion
Wallet Instance. In the non-normative example below is represented an HTTP Response with the `status_assertion_responses` JSON member:
```

~~~
HTTP/1.1 201 Created
Content-Type: application/json

{
    "status_assertion_responses": ["${base64url(json({typ: status-assertion+jwt, ...}))}.payload.signature", ... ]
}
~~~

The member `status_assertion_responses` MUST be an array of strings,

> where each of them represent a Status Assertion Response object and the following requirements are met:

Reviewer: ```suggestion
where each of them represent a Status Assertion Response object or a Status Assertion Error object.
For each entry in the `status_assertion_responses` array, the following requirements are met:
```

> - Each element in the array MUST match the corresponding element in the request array at the same index

Reviewer: ```suggestion
- Each element in the array MUST match the corresponding element in the request array at the same index to which it is related.
```

> - Each element MUST contain the error or the status of the assertion using the `typ` member

Reviewer: ```suggestion
- Each element MUST contain the error or the status of the assertion using the `typ` member.
```

> set to "status-assertion-error+{jwt,cwt}" or "status-assertion+{jwt,cwt}"

Reviewer: ```suggestion
set to "status-assertion-error+{jwt,cwt}" or "status-assertion+{jwt,cwt}", depending by the object type.
```

If the Digital Credential is valid, the Credential Issuer MUST return
an HTTP status code of 201 (Created), with the content type set to
`application/json`. The response MUST include a JSON object with a member
named `status_assertion_responses`, which contains the Status Assertion for the
Wallet Instance, as illustrated in the following non-normative example:

~~~
HTTP/1.1 201 Created
Content-Type: application/json

{
    "status_assertion_responses": ["${base64url(json({typ: status-assertion+jwt, ...}))}.payload.signature", ... ]
}
~~~

The member `status_assertion_responses` MUST be an array of strings,

> where each of them represent a Status Assertion Response object. In particular:
> to the following specification:

Reviewer: ```suggestion
where each of them represent a Status Assertion Response object. In particular:
```

> - Each element in the array MUST match the element contained in the request at the same position

Reviewer: at the same position ... ? of what.

Reviewer 3: ```suggestion
- Each element in the array MUST match the corresponding element in the request array at the same index
```

- Each element MUST contain the error or the status of the assertion using the `typ` member

> set to "status-assertion-error+{jwt,cwt}"
> or "status-assertion+{jwt,cwt}"

Reviewer: ```suggestion
set to "status-assertion-error+{jwt,cwt}" or "status-assertion+{jwt,cwt}"
```

> - The corresponding entry in the response must be of the same type as requested. For example,

Reviewer: ```suggestion
- The corresponding entry in the response MUST be of the same type as requested. For example,
```
